cpp/src/arrow/filesystem/hdfs.cc

```diff
@@ -0,0 +1,298 @@
+// specific language governing permissions and limitations
+// under the License.
+
+#include <chrono>
+#include <cstring>
+#include <utility>
+
+#ifdef _WIN32
+#include "arrow/util/windows_compatibility.h"
+#else
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#endif  // _WIN32
+
+#include <boost/filesystem.hpp>
+
+#include "arrow/filesystem/hdfs.h"
+#include "arrow/filesystem/util_internal.h"
+#include "arrow/io/hdfs.h"
+#include "arrow/io/hdfs_internal.h"
+#include "arrow/util/io_util.h"
+#include "arrow/util/logging.h"
+
+namespace arrow {
+namespace fs {
+
+class HadoopFileSystem::Impl {
+ public:
+  explicit Impl(HdfsOptions options) : options_(std::move(options)) {}
+  ~Impl() { auto status = Close(); }
+
+  Status Init() {
+    io::internal::LibHdfsShim* driver_shim;
+    if (options_.hdfs_config.driver == io::HdfsDriver::LIBHDFS3) {
+      RETURN_NOT_OK(ConnectLibHdfs3(&driver_shim));
+    } else {
+      RETURN_NOT_OK(ConnectLibHdfs(&driver_shim));
+    }
+    RETURN_NOT_OK(io::HadoopFileSystem::Connect(&options_.hdfs_config, &client_));
+    return Status::OK();
+  }
+
+  Status Close() {
+    if (client_) {
+      RETURN_NOT_OK(client_->Disconnect());
+    }
+    return Status::OK();
+  }
+
+  Status GetTargetStats(const std::string& path, FileStats* out) {
+    io::HdfsPathInfo info;
+    RETURN_NOT_OK(client_->GetPathInfo(path, &info));
+
+    out->set_path(path);
+    if (info.kind == io::ObjectType::DIRECTORY) {
+      out->set_type(FileType::Directory);
+      out->set_size(kNoSize);
+    } else {
+      out->set_type(FileType::File);
+      out->set_size(info.size);
+    }
+    out->set_mtime(ToTimePoint(info.last_modified_time));
+    return Status::OK();
+  }
+
+  Status GetTargetStats(const Selector& select, std::vector<FileStats>* out) {
+    std::vector<std::string> file_list;
+    RETURN_NOT_OK(client_->GetChildren(select.base_dir, &file_list));
+    for (auto file : file_list) {
+      FileStats stat;
+      RETURN_NOT_OK(GetTargetStats(file, &stat));
+      out->push_back(stat);
+    }
+    return Status::OK();
+  }
+
+  Status CreateDir(const std::string& path, bool recursive) {
+    if (client_->Exists(path)) {
+      return Status::OK();
+    }
+    RETURN_NOT_OK(client_->MakeDirectory(path));
+    return Status::OK();
+  }
+
+  Status DeleteDir(const std::string& path) {
+    if (!IsDirectory(path)) {
+      return Status::IOError("path is not a directory");
+    }
+    if (!client_->Exists(path)) {
+      return Status::OK();
+    }
+    RETURN_NOT_OK(client_->DeleteDirectory(path));
+    return Status::OK();
+  }
+
+  Status DeleteDirContents(const std::string& path) {
+    if (!IsDirectory(path)) {
+      return Status::IOError("path is not a directory");
+    }
+    std::vector<std::string> file_list;
+    RETURN_NOT_OK(client_->GetChildren(path, &file_list));
+    for (auto file : file_list) {
+      RETURN_NOT_OK(client_->Delete(path + file));
```

Comment:
What if `file` is a directory? `DeleteDirContents` is supposed to be recursive.

Reply:
Delete Function in io/hdfs.cc isable to delete the input path either is file or directory, and I will change the recursive flag to true.